### DISCUSSION TITLE
A little help with OpenGL

### DISCUSSION THREAD

**Discussion by mardy** [Neverball/neverball#388] [Q&A] [answered]
Hi! I'm making good progress with the Wii port of Neverball, most things are rendering fine now, but there's one bit of code that I had to change in neverball's source code, which puzzles me a bit.

It's about the ball's shadow: unless I comment out the condition at https://github.com/Neverball/neverball/blob/master/share/geom.c#L597 (that is, unless I explicitly set the wrapping mode to "clamp"), the ball shadow texture is repeated all over the stage. I'm using the OpenGL driver I'm developing myself, so it's very likely that the bug is somewhere in there, but I don't understand how can the code work on the desktop, since, AFAIK, the default value for the texture wrap mode is GL_REPEAT. Can someone explain me why the ball shadow is not repeated, on proper OpenGL drivers?

For the record, I'm not using the latest master branch of Neverball, but the same 2017 version packaged in Ubuntu (just temporarily, in order to be able to reuse the data files :-p).

**Comment by parasti** [comment]
Hi,

First of all, a Wii port sounds awesome.

Second, you must be doing more than you're letting on, because in Neverball code `shad_init` calls `make_image_from_file` which calls `make_texture`, which does indeed explicitly set WRAP modes to CLAMP_TO_EDGE. https://github.com/Neverball/neverball/blob/8d01ce92fe23c13879a059624852eb2d2dcc4cc9/share/image.c#L139

**Reply by mardy** [reply]
Thanks, you are indeed correct: I had an error in my opengx driver, where glTexImage2D was resetting any wrapping modes earlier set.

I've still one big issue related to shadows (at least I think it's related, as the problem disappears if shadows are disabled) which only happens on non-flat floors. I'm attaching a short video I recorded on the Dolphin emulator, which shows some black areas, which seem to be related to the ball's height from the floor. I think it might be related with some errors on how I process the clip texture, and the main obstacle in debugging it is that it's not clear to me what the clip texture unit does. Maybe if you explain it, I can get a hint on what I could be doing wrong?

https://github.com/user-attachments/assets/db77a306-98ed-461d-8e24-5da93dba4c86

(on a real Wii the FPS are way higher, about 60 I believe)

**Reply by mardy** [reply]
Or, to put it more practically: in which level/situation can I see the clip texture in action? (like, if I disable it, what actually breaks? Because I cannot seem to notice any difference)

**Reply by parasti** [reply, marked as answer]
The clip texture serves one purpose: to clip the shadow above the ball. This was so common in Neverball levels that it actually had a FAQ entry on the old Neverball website. And so I fixed it. The fix was to combine the shadow texture with a clip texture that is positioned such that the shadow color above the midpoint of the ball is multiplied with an alpha value of 0 - and disappears.

In your video, the clip texture is positioned correctly, but appears to be blended incorrectly: it should only affect the shadow color and should itself be invisible.

(I guess there's no difference from using a clip texture vs using a clipping plane - I just did it with a texture.)

**Reply by mardy** [reply]
> The clip texture serves one purpose: to clip the shadow above the ball. This was so common in Neverball levels that it actually had a FAQ entry on the old Neverball website.

I think I understand the fix, but not really the problem: under what scenarios can the shadow extend above the ball? I disabled the 2nd texture unit (the clipping one), but I cannot seem to notice any difference; can you please suggest a level where I should see the problem?

> In your video, the clip texture is positioned correctly, but appears to be blended incorrectly: it should only affect the shadow color and should itself be invisible.

Mmm... yes, it may be that I'm not picking the correct value for GL_PRIMARY_COLOR in the GL_INTERPOLATE operation.

**Reply by parasti** [reply]
You can see it whenever the ball goes under a platform. E.g., Neverball Hard first bonus level.

**Reply by mardy** [reply]
Found it, thanks! And I think I found a way to correct my OpenGL implemenation :-)
